app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx

```diff
@@ -0,0 +1,941 @@
+/** @file A WebSocket-based chat directly to official support on the official Discord server. */
+import * as react from 'react'
+import * as reactDom from 'react-dom'
+import toast from 'react-hot-toast'
+
+import CloseLargeIcon from 'enso-assets/close_large.svg'
+import DefaultUserIcon from 'enso-assets/default_user.svg'
+import TriangleDownIcon from 'enso-assets/triangle_down.svg'
+
+import * as animations from '../../animations'
+import * as authProvider from '../../authentication/providers/auth'
+import * as dateTime from '../dateTime'
+import * as loggerProvider from '../../providers/logger'
+import * as newtype from '../../newtype'
+
+import Twemoji from './twemoji'
+
+// =================
+// === Constants ===
+// =================
+
+// TODO[sb]: Consider associating a project with a thread
+// (and providing a button to jump to the relevant project).
+// The project shouldn't be jumped to automatically, since it may take a long time
+// to switch projects, and undo history may be lost.
+
+const HELP_CHAT_ID = 'enso-chat'
+export const ANIMATION_DURATION_MS = 200
+const WIDTH_PX = 352
+/** The size (both width and height) of each reaction button. */
+const REACTION_BUTTON_SIZE = 20
+/** The size (both width and height) of each reaction on a message. */
+const REACTION_SIZE = 16
+/** The list of reaction emojis, in order. */
+const REACTION_EMOJIS: ReactionSymbol[] = ['❤️', '👍', '👎', '😀', '🙁', '👀', '🎉']
+/** The initial title of the thread. */
+const DEFAULT_THREAD_TITLE = 'New chat thread'
+/** The maximum number of lines to show in the message input, past which a scrollbar is shown. */
+const MAX_MESSAGE_INPUT_LINES = 10
+/** The maximum number of messages to fetch when opening a new thread.
+ * This SHOULD be the same limit as the chat backend (the maximum number of messages sent in
+ * `serverThread` events). */
+const MAX_MESSAGE_HISTORY = 25
+
+// =============
+// === Types ===
+// =============
+
+/** All possible reaction emojis. */
+type ReactionSymbol = '❤️' | '🎉' | '👀' | '👍' | '👎' | '😀' | '🙁'
+
+// =====================
+// === Message Types ===
+// =====================
+
+// FIXME[sb]: Consider deduplicating.
```

Comment:
This should be part of cleaning this PR if its about code deduplication :) if its about anything else, it needs better docs.